refactor: update public facing APIs to be more pythonic

ajbozarth opened the pull request:
### Which issue(s) does this pull-request address?

Closes: #385

### Description

In our current implementation we leverage BaseModel input objects for our various APIs, this is really great for clear typing and passing of data between functions, but it is not very user friendly for our public-facing APIs.

In this PR we update our public-facing APIs to "spread" the BaseModel input params and update any references and examples.

#### Work Items
- [x] `BaseAgent.run`
- [ ] `BeeAgent.__init__`
- [ ] `ChatModel.create`
- [ ] `ChatModel.create_structure`

Open discussion work items: https://github.com/i-am-bee/beeai-framework/issues/385#issuecomment-2683307173

### Checklist

- [x] I have read the [contributor guide](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md)
- [x] I have [signed off](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-dco) on my commit
- [ ] Tests are included <!-- Bug fixes and new features should include tests -->
- [x] Documentation is changed or added
- [x] Commit messages and PR title follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)

ajbozarth commented:
Moved out of draft as the agent updates could be merged as is and don't need to be blocked by my continued work on the ChatModel updates

ajbozarth commented:
A small note on my `update ChatModel.create API` commit: I was unable to test the `watsonx.py` and `web_agent.py` examples due to not having WatsonX or SearX running locally

ajbozarth commented:
Just pushed the last commit for the core work items, I will note that my IDE kept clearing the notebook output with my last update so this might require a followup PR by @michael-desmond to address that.

ajbozarth commented on python/beeai_framework/agents/base.py:
```diff
@@ -31,8 +31,8 @@ class BaseAgent(ABC, Generic[T]):
 
-    def run(self, run_input: ModelLike[BeeRunInput], options: ModelLike[BeeRunOptions] | None = None) -> Run[T]:
-        run_input = to_model(BeeRunInput, run_input)
+    def run(self, prompt: str | None = None, options: ModelLike[BeeRunOptions] | None = None) -> Run[T]:
+        run_input = BeeRunInput(prompt=prompt)
```
I don't personally understand the use case for it, but that's how TS is written and it is called that way in `workflows/agent.py` so I just made sure to support it.

Tomas2D replied:
tldr: ability to run an agent without adding a new message

- typical use case for APIs, where memory is managed externally
- re-run the agent after some changes (tool changes, implementation changes)

ajbozarth commented:
I've pushed an update that merges in latest main and addresses relevant changes in main. I'll be making the updates to Emitter next

Tomas2D commented on python/beeai_framework/backend/chat.py:
```diff
@@ -215,25 +213,40 @@ async def executor(_: RetryableContext) -> ChatModelStructureOutput:
+    def create(
+        self,
+        messages: list[Message],
+        tools: list[Message] | None = None,
```
Use `*` to force keyword arguments to preserve backward compatibility if we change the order in the future

Tomas2D commented on python/beeai_framework/backend/chat.py:
```diff
@@ -246,19 +259,27 @@ async def run_create(context: RunContext) -> ChatModelOutput:
-        input = to_model(ChatModelStructureInput, structure_input)
+    def create_structure(
+        self,
+        schema: type[T],
```
same here

Tomas2D commented on python/beeai_framework/memory/summarize_memory.py:
```diff
@@ -61,7 +61,7 @@ async def _summarize_messages(self, messages: list[Message]) -> str:
         )
 
         # Generate is synchronous, not async
```
we can remove this comment 😄

ajbozarth commented:
@jenna-winkler I added some doc updates here, I almost didn't notice but a few of our doc code blocks are "based on" examples, but not embedded. I've checked and made sure all the docs reflect this API update.

@Tomas2D I've also rebased on you updates from today so this is safe for auto-merge again

Tomas2D commented on python/docs/agents.md:
```diff
@@ -269,16 +268,14 @@ async def main() -> None:
-                },
-            )
+            prompt=user_input,
+            options={
+                "execution": {
+                    "max_retries_per_step": 3,
+                    "total_max_retries": 10,
+                    "max_iterations": 20,
+                }
+            },
```
maybe we could flatten this to just

```
await agent.run(prompt=user_input, execution={"..."})
# or
await agent.run(prompt=user_input, execution=BeeExecution(...)) or 
```

ajbozarth replied:
This is a embedded example, so this update comes from `examples/memory/agentMemory.py`

ajbozarth replied:
I can update it there, but I think it's good to show multiple ways to pass in options since we support it

Tomas2D replied:
I was thinking about changing the signature of the `run` method

ajbozarth replied:
oh, options actually has two value in it though: execution and signal, also the options object is passed through to _run and eventually the runner, but I can still flatten both those value into run and create the internal options object like I did elsewhere. I'll add it to #430

ajbozarth commented:
I'll open a followup PR to address feedback and small conflict with #416
